mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -2319,3 +2374,148 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
+                  if (sdInfo.stride * sdInfo.size >= maxStrideSubDim) {
+                    int64_t rhsBoundForAnnotate =
+                        std::max(sdInfo.stride, subDimStrides[subDim]);
+                    newSize = maxStrideSubDim / rhsBoundForAnnotate;
+                    newStride = rhsBoundForAnnotate / subDimStrides[subDim];
+                  }
+                  // The whole of annotatedSubDim is within the newly created
+                  // lowDim
+                  else if (sdInfo.stride >= subDimStrides[subDim]) {
+                    newSize = sdInfo.size;
+                    newStride = sdInfo.stride / subDimStrides[subDim];
+                  }
+                  // Overlap on the left side of annotated subdim
+                  else {
+                    int64_t maxStrideRemovedSubDim =
+                        sdInfo.stride * sdInfo.size;
+                    newSize = maxStrideRemovedSubDim / subDimStrides[subDim];
+                    newStride = 1;
+                  }
+                  LLVM_DEBUG(llvm::dbgs() << "creating subDim of <size="
+                                          << newSize << ",stride=" << newStride
+                                          << "> @ " << lowDim << "\n");
+                  if (newSize > 1) {
+                    nextSubDimInfo[lowDim].push_back({newSize, newStride});
+                  }
+                }
+              }
+            }
+          }
+        } break;
+        case TransformType::Unmerge: {
+          SmallVector<int64_t> subDimStrides = getStrides(trAttr.getParams());
+          int64_t lowDim = trAttr.getLowerDims()[0];
+          for (size_t subDim = 0; subDim < trAttr.getParams().size();
+               subDim++) {
+            int64_t upperDim = trAttr.getUpperDims()[subDim];
+            if (currSubDimInfo.contains(upperDim)) {
+              for (const SubDimInfo &sdInfo : currSubDimInfo.at(upperDim)) {
+                int64_t newStride = sdInfo.stride * subDimStrides[subDim];
+                LLVM_DEBUG(llvm::dbgs()
+                           << "creating subDim of <size=" << sdInfo.size
+                           << ",stride=" << newStride << "> @ " << lowDim
+                           << "\n");
+                if (sdInfo.size > 1) {
+                  nextSubDimInfo[lowDim].push_back({sdInfo.size, newStride});
+                }
+              }
+            }
+          }
+          break;
+        }
+        default:
+          LLVM_DEBUG(llvm::dbgs()
+                     << "Unsupported transform type : " << trAttr << "\n");
+          return failure();
+        }
+      }
+      return nextSubDimInfo;
+    };
+    FailureOr<llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>>>
+        nextSubDimInfo = getNextSubDimInfo(subDimInfo);
+    if (failed(nextSubDimInfo))
+      return failure();
+    subDimInfo = nextSubDimInfo.value();
+  }
+  if (subDimInfo.empty()) {
+    return failure();
+  }
+  return subDimInfo;
+}
```

Comment:
Out of curiosity, could this SubDimInfo stuff be used to simplify `collapseContiguousMerges()`?

(Not for this PR, just checking my understanding)

Reply:
potentially yes...